### DISCUSSION TITLE
Test if internal linked to page exists

### DISCUSSION THREAD

**Discussion by Martien** [observablehq/framework#1858] [Q&A] [answered]
In order to avoid broken links, I need to test if a linked to internal page exists and not link to it if it doesn’t, yet keep the text and style it accordingly. That is, omit the `href` in `<a>`.

Need it both for Markdown files and Page Loaders.

Is there an existing helper function in Framework that I can use?

Context: I am generating (a lot of) page names and corresponding links that I know will eventually exist (but not yet). `npm run build` lists me all non-existing references in red, and I want to get rid of them.

**Comment by Fil** [comment, marked as answer]
> Is there an existing helper function in Framework that I can use?

Short answer is no. The internal function that does this is the `findPage` method on the `LoaderResolver` class, but it is not  publicly exposed or available in user land.

For markdown pages, it's quite simple to emulate (just check if the corresponding md file is available); but it's a bit harder if you want to include dynamic paths and page loaders, since the results depend on the configuration (`interpreters`). In addition, you might want to check if the page's front-matter includes the `draft` key.

An alternative possibility, that might be simpler in the end, is to build the whole site, then use the generated files to check whether a given page exists. Or, you could also use the search index (`minisearch.json`), where the `documentIds` key corresponds to a list of all the existing paths (for indexed pages).

Or, we could have an option to remove the `href` from so-called red links (https://github.com/observablehq/framework/discussions/1828, https://github.com/observablehq/framework/issues/1730).

**Reply by Martien** [reply]
Thanks for your time and effort, Fil.

Martien – tucked away in a corner, crying.
